src/System.IO.Pipelines/BufferSegment.cs

```diff
@@ -91,11 +91,6 @@ public void Dispose()
             Debug.Assert(_owned.HasOutstandingReferences);
 
             _owned.Release();
-
-            if (!_owned.HasOutstandingReferences)
```

Comment:
Here is the race. If BufferHandle.Free() releases the last reference to the _owned buffer, _owned would never get disposed. This is fixed by always returning MemoryPoolBlocks to the pool in OnZeroReferences.